talk/objectorientation/adl.tex

```diff
@@ -168,3 +173,62 @@
     \end{cppcode*}
   \end{exampleblock}
 \end{frame}
+
+\begin{frame}[fragile]
+  \frametitlecpp[98]{Hidden Friends and ADL}
+  \begin{block}{Advantages of hidden friends}
+    \begin{itemize}
+      \item Are \emph{only} found via ADL, e.g.\ \mintinline{cpp}{std::cout << complex;}
+      \item Compiler needs to consider less functions during ordinary name lookup (faster compilation)
+      \item Avoid accidental implicit conversions
+    \end{itemize}
+  \end{block}
+  \begin{alertblock}{Accidental conversion example}
+    \small
+    \begin{cppcode*}{gobble=2}
+      std::ostream & operator<<(std::ostream & os,
+        Complex const & c) { ... } // out-of-class definition
+      // either: Complex::Complex(Fraction f) { ... }
+      struct Fraction {
+        int num, denom;
+        operator Complex() const { ... } // or: conversion op
+      };
+      std::cout << Fraction{2, 3}; // Prints 0.6666 + 0i
+      // operator<<(std::cout, Complex(Fraction{2, 3}));
+      // operator<<(std::cout, Fraction{2, 3}.operator Complex());
```

Comment:
Why are these lines commented ? They would compile. So strange things, but compile

Reply:
Because these lines show what would we called in the two cases. I will make that more clear.